lib/ThreeElementVectorLoweringPass.cpp

```diff
@@ -294,24 +275,50 @@ bool vec3ShouldBeLowered(Module &M) {
   }
 }
 
-} // namespace
+bool clspv::ThreeElementVectorLoweringPass::vec3BitcastInFunction(Function &F) {
+  for (Instruction &I : instructions(F)) {
+    if (auto *Inst = dyn_cast<CastInst>(&I)) {
+      auto *Type = Inst->getSrcTy();
+      if (Type->isPointerTy() && !Type->isOpaquePointerTy()) {
+        auto *PointeeType = Type->getNonOpaquePointerElementType();
+        if (auto *VectorType = dyn_cast<FixedVectorType>(PointeeType)) {
+          if (VectorType->getElementCount().getKnownMinValue() == 3)
+            return true;
+        }
+      }
+    }
+  }
 
-PreservedAnalyses
-clspv::ThreeElementVectorLoweringPass::run(Module &M, ModuleAnalysisManager &) {
-  PreservedAnalyses PA;
-  if (!vec3ShouldBeLowered(M))
-    return PA;
+  int ArgIndex = 0;
+  for (auto &Arg : F.args()) {
+    // If the argument is of opaque pointer type...
+    if (Arg.getType()->isOpaquePointerTy() &&
+        F.getMetadata("kernel_arg_type")) {
+      auto const &type_op =
+          F.getMetadata("kernel_arg_type")->getOperand(ArgIndex);
+      auto const &type_name_str = dyn_cast<MDString>(type_op)->getString();
 
-  runOnGlobals(M);
-  for (auto &F : M.functions()) {
-    BitcastUtils::RemovedCstExprFromFunction(&F);
-    runOnFunction(F);
-  }
+      auto ArgInferredType =
+          clspv::InferType(&Arg, F.getContext(), &type_cache_);
 
-  cleanDeadFunctions();
-  cleanDeadGlobals();
+      // ... and its main type is a vec3 type...
+      if (type_name_str.endswith("3*")) {
+        // ... and its inferred type is of scalar type...
+        if (!ArgInferredType->isVectorTy()) {
+          return true;
+        }
 
-  return PA;
+        // .. or have a mismatched vector size...
+        auto *VectorType = dyn_cast<FixedVectorType>(ArgInferredType);
+        if (VectorType->getElementCount().getKnownMinValue() != 3) {
+          return true;
+        }
+      }
+    }
```

Comment:
Do we better have this like that or just checking if the metadata or inferred type is a vector of size 3 would be simpler and would still work?

Reply:
I'd prefer not to rely on metadata. We want to identify cases where a vec3 buffer is being used as a non-vec3. If InferType returns returns a vec3 that is subsequently treated as a different type that's what we need to catch. So probably we want to catch where the buffers are inferred as vec3, but then there is an implicit cast to a non-vec3 elsewhere in the flow.

For example the following is perfectly ok:
```
kernel void foo(global float3* data) {
  global float* tmp = (global float*)data;
  *data = 0;
}
```
clspv will just treat this as a float buffer, but:
```
kernel void bar(global float3* data) {
  global float3* tmp1 = data[3];
  global float4* tmp2 = (global float4*)tmp1;
  *tmp2 = (float4)(0.0f,0.0f,0.0f,0.0f);
}
```
This won't be caught (in the naive translation at least).

Reply:
Updated, I have used the stores to detect implicit casts.